File: spaced_rep_java_oct_2024/coin_change.java
package spaced_rep_java_oct_2024;

import java.util.Arrays;

class Solution {
    // Bottom-Up Dynamic Programming.
    // Time and Space: O(n * m)...n for coins and m for amount
    // Space: O(m)...m for amount
    public int coinChange(int[] coins, int amount) {
        int dp[] = new int[amount+1];
        Arrays.fill(dp, amount+1);
        dp[0] = 0;

        for (int c : coins) {
            for (int a = 1; a <= amount; a++) {
                if (a - c >= 0) {
                    dp[a] = Math.min(dp[a], dp[a-c] + 1);
                }
            }
         }
        return dp[amount] == amount+1 ? -1 : dp[amount];
    }
    //   coins = [1,2,5], amount = 11, Result = 3
}

File: spaced_rep_nov_2024 (1)/coinChange.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */

// Bottom-Up Dynamic Programming
// Time: O(n * m)...n for # of coins and m for amount
// Space: O(m)
var coinChange = function(coins, amount) {
    const dp = [...new Array(amount+1)].fill(Infinity)
    dp[0] = 0;
    for (let c of coins) {
        for (let a = 1; a <= amount; a++) {
            if (a - c >= 0) {
                dp[a] = Math.min(dp[a], 1 + dp[a - c])
            }
        }
    }
    return dp[amount] === Infinity ? -1 : dp[amount]
};
console.log(coinChange(coins = [1,2,5], amount = 11))
// = 1 + 5 + 5 = (3 coins)

